sdk/textanalytics/ai-text-analytics/src/analyzeBatchActionsResult.ts

```diff
@@ -47,3 +155,312 @@ export interface PagedAnalyzeBatchActionsResult
+ * @param erredActions - the action error the service sent
+ * @returns an action error with an action type and index
+ * @internal
+ */
+export function parseActionError(erredActions: TextAnalyticsError): TextAnalyticsActionError {
+  if (erredActions.target) {
+    const regex = new RegExp(
+      /#\/tasks\/(entityRecognitionTasks|entityRecognitionPiiTasks|keyPhraseExtractionTasks)\/(\d+)/
+    );
+    const res = regex.exec(erredActions.target);
+    if (res !== null) {
+      return {
+        code: erredActions.code,
+        message: erredActions.message,
+        index: parseInt(res[2]),
+        type: convertTaskTypeToActionType(res[1])
+      };
+    } else {
+      throw new Error(`Pointer "${erredActions.target}" is not a valid action pointer`);
+    }
+  } else {
+    throw new Error(
+      "expected an error with a target field referencing an action but did not get one"
+    );
+  }
+}
+
+/**
+ * Categorize each action error into a bucket according to its action type.
+ * @param erredActions - list of action errors
+ * @param recognizeEntitiesActionErrors - a list of recognize entities action errors to be filled from the errors list
+ * @param recognizePiiEntitiesActionErrors - a list of recognize pii entities action errors to be filled from the errors list
+ * @param extractKeyPhrasesActionErrors - a list of extract key phrases action errors to be filled from the errors list
+ * @internal
+ */
+function categorizeActionErrors(
+  erredActions: TextAnalyticsError[],
+  recognizeEntitiesActionErrors: TextAnalyticsActionError[],
+  recognizePiiEntitiesActionErrors: TextAnalyticsActionError[],
+  extractKeyPhrasesActionErrors: TextAnalyticsActionError[]
+): void {
+  for (const error of erredActions) {
+    const actionError = parseActionError(error);
+    switch (actionError.type) {
+      case "RecognizeCategorizedEntities": {
+        recognizeEntitiesActionErrors.push(actionError);
+        break;
+      }
+      case "RecognizePiiEntities": {
+        recognizePiiEntitiesActionErrors.push(actionError);
+        break;
+      }
+      case "ExtractKeyPhrases": {
+        extractKeyPhrasesActionErrors.push(actionError);
+        break;
+      }
+      default: {
+        throw new Error(`Expected a recognized action type but got ${actionError.type}`);
+      }
+    }
+  }
+}
+
+/**
+ * Combines the lists of succeeded and erred actions into one that is ordered with respect to the input actions list.
+ * @param succeededActions - a list of succeeded actions
+ * @param erredActions - a list of erred actions
+ * @returns a list of succeeded and erred actions
+ * @internal
+ */
+export function combineSucceededAndErredActions<TSuccess extends TextAnalyticsActionSuccessState>(
+  succeededActions: TSuccess[],
+  erredActions: TextAnalyticsActionError[]
+): (TSuccess | TextAnalyticsActionErrorResult)[] {
+  const actions: (TSuccess | TextAnalyticsActionErrorResult)[] = [];
+  for (
```

Comment:
One ask, one nit

Ask: Is it possible to declare errorIndex separately? It doesn't get incremented as part of the iteration (only in the body) so it feels strange to initialize it here

Nit: If it's all the same to you would it be possible to move the incrementing to the right hand side? Since you aren't using the result of the increment it doesn't matter which side is used on and `actionIndex++` at least to me is more common. Feel free to ignore of course if you have a strong preference or this is more consistent!

Reply:
> Ask: Is it possible to declare errorIndex separately? It doesn't get incremented as part of the iteration (only in the body) so it feels strange to initialize it here

It is possible of course, I just wanted to limit its scope to just where it is actually used.

> If it's all the same to you would it be possible to move the incrementing to the right hand side?

Yeah it is more common. I prefer it this way though because it updates the incremented variable right away and no need to keep a copy of the old value around in case it will be used. You can see it is compiled to fewer instructions by trying  `int x = ++num;` in https://godbolt.org/. Forgive me for being this pedantic for no good reason :)